Implement command to validate builds

JenySadadia opened the pull request:
Closes https://github.com/kernelci/kci-dev/issues/171

Get builds from dashboard and validate it with maestro result.
PR created on top of https://github.com/kernelci/kci-dev/pull/177 and https://github.com/kernelci/kci-dev/pull/179
Depends on https://github.com/kernelci/kernelci-api/pull/615

padovan commented:
I am getting this when running the command without arguments:

```
kci-dev > poetry run kci-dev validate-builds
Validate builds called----
git folder: None
tree: https://github.com/kernelci/kci-dev.git
branch: validate-builds-cmd
commit: 99f9ec1a74f6d6817feae8902f627187a997cdf6
json error: Tree checkout not found
Aborted!
```

Guess we need to throw a clear error message guiding the user.

JenySadadia commented:
> I am getting this when running the command without arguments:
> 
> ```
> kci-dev > poetry run kci-dev validate-builds
> Validate builds called----
> git folder: None
> tree: https://github.com/kernelci/kci-dev.git
> branch: validate-builds-cmd
> commit: 99f9ec1a74f6d6817feae8902f627187a997cdf6
> json error: Tree checkout not found
> Aborted!
> ```
> 
> Guess we need to throw a clear error message guiding the user.

Thanks. Addressed this.

JenySadadia commented:
For starters, the command will work by providing git URL, branch, and commit.
Here are the samples:
```
$ kci-dev --settings ./kci-dev.toml validate-builds --giturl https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git --branch master --commit 7d4e49a77d9930c69751b9192448fda6ff9100f1 --arch arm
Maestro:
Total nodes found in maestro:  15
Dashboard:
15
Build count matched
```
```
$ kci-dev --settings ./kci-dev.toml validate-builds --commit fb01b86bf35a2aecfb4e777c1cb81a69203edc0b --giturl https://github.com/kernelci/linux.git --branch staging-next
Maestro:
Total nodes found in maestro:  0
Dashboard:
json error: Tree checkout not found
Aborted!
```

JenySadadia commented:
Modified directory structure, refactored the code and commands.
The new commands are as below:

1. `kci-dev validate build all-checkouts`
2. `kci-dev validate build specific-checkout --giturl <URL> --commit <commit-hash> --branch <git-branch>`

Both commands will provide below information:
1. Build count validation
2. Missing build information
3. Build status validation

aliceinwire commented:
When using `specific-checkout` the command should get information from the current folder, in a similar fashion of `kci-dev results summary` for example

JenySadadia commented:
> Thanks for this, @JenySadadia.
> 
> As for the overall command structure. I have been thinking that we need to create `kci-dev maestro`, so we can move a lot of maestro specific commands under it. (eg `kci-dev maestro-results` should become `kci-dev maestro results`).
> 
> So I suggest:
> 
> 1. A small detour to start ``kci-dev maestro`adding`kci-dev maestro results` under it.
> 2. Make this patchset use that: `kci-dev maestro validate`
> 
Yes, makes sense.
I'll create another PR for this as this is out of scope for the current PR.

> Now looking at the options you added, my suggestion would be:
> 
> `kci-dev maestro validate build --all-checkouts` `kci-dev maestro validate build --giturl <URL> --commit <commit-hash> --branch <git-branch>`
> 
> Let's also make the output a bit more standard:
> 
> ```
> Maestro:  45
> Dashboard: 45
> Build count matched
> ```

Thanks for the suggestion. Addressed this.
The output looks like the below now:
```
Maestro:
Build count: 45
Dashboard:
Build count:45
Build count matched
```

padovan commented:
> When using `specific-checkout` the command should auto fill missing parameters, in a similar fashion of `kci-dev results summary` for example

That is not so important for this usecase. validate-builds will be more a of backend validation tool for maestro, so not the usual kernel dev doing his development.

aliceinwire commented:
> > When using `specific-checkout` the command should auto fill missing parameters, in a similar fashion of `kci-dev results summary` for example
> 
> That is not so important for this usecase. validate-builds will be more a of backend validation tool for maestro, so not the usual kernel dev doing his development.

I had found it useful when testing the command

aliceinwire commented:
Are we supposed to see `Missing builds:` as a Json dictionary string ? or we can show it in a better way?

```
Missing builds: [{'id': '683f98f5214050d3418d0917', 'kind': 'kbuild', 'name': 'kbuild-clang-17-x86-chromeos-daily-intel', 'path': ['checkout', 'kbuild-clang-17-x86-chromeos-daily-intel'], 'group': 'kbuild-clang-17-x86-chromeos-daily-intel', 'parent': '683f9836214050d3418d03a5', 'state': 'done', 'result': 'incomplete', 'artifacts': {}, 'data': {'kernel_revision': {'tree': 'chromiumos', 'url': 'https://chromium.googlesource.com/chromiumos/third_party/kernel.git', 'branch': 'chromeos-6.1', 'commit': 'e05128ed2715843e6e03ce05d20ca32cb82a493e', 'describe': 'v6.1.129-17451-ge05128ed27158', 'version': {'version': 6, 'patchlevel': 1, 'sublevel': 129, 'extra': '-17451-ge05128ed27158'}, 'commit_tags': [], 'commit_message': 'CHROMIUM: config: Enable CONFIG_USB_LINK_LAYER_TEST for MTK\n\nEnable CONFIG_USB_LINK_LAYER_TEST as a module to add the lvstest\ndriver.\n\nBUG=b:412565469\nTEST=CQ, tested on 6.6 kernel\n\nChange-Id: Icbdc91eb35253a925a1a03bbadffd87f7e89c2c2\nSigned-off-by: Jameson Thies <jthies@google.com>\nReviewed-on: https://chromium-review.googlesource.com/c/chromiumos/third_party/kernel/+/6594756\nReviewed-by: Abe Levkoy <alevkoy@chromium.org>', 'tip_of_branch': True}, 'runtime': 'k8s-all', 'platform': 'kubernetes', 'job_id': 'kci-683f98f5214050d3418d0917-kbuild-clang-17-x86-chro-t45qnxb0', 'job_context': 'aks-kbuild-medium-1', 'error_code': 'node_timeout', 'error_msg': 'Node timed-out'}, 'debug': {}, 'jobfilter': None, 'platform_filter': [], 'created': '2025-06-04T00:53:09.138000', 'updated': '2025-06-04T07:02:53.365000', 'timeout': '2025-06-04T06:53:09.138000', 'holdoff': None, 'owner': 'production', 'submitter': 'service:pipeline', 'treeid': '8e05534a68f0a8d700c47ec16323ebab2f5d25a3eda2f4fbcacb97f6c85da43a', 'user_groups': [], 'processed_by_kcidb_bridge': True}]
```

JenySadadia commented:
@aliceinwire I reformatted output to be nicer:
```
$ kci-dev --settings ./kci-dev.toml validate build --giturl https://android.googlesource.com/kernel/common --commit de3036f041b742dfd2b0d86fc50ce61aaeca0aa1 --branch android12-5.4
Maestro:
Build count:34
Dashboard:
Build count:33
Build count didn't match
[
 {
  "id": "6849fb29dba9af5ad62c489b",
  "kind": "kbuild",
  "name": "kbuild-clang-17-x86-android-allmodconfig",
  "path": [
   "checkout",
   "kbuild-clang-17-x86-android-allmodconfig"
  ],
  "group": "kbuild-clang-17-x86-android-allmodconfig",
  "parent": "6849f8c8dba9af5ad62c4753",
  "state": "done",
  "result": "incomplete",
  "artifacts": {},
  "data": {
   "kernel_revision": {
    "tree": "android",
    "url": "https://android.googlesource.com/kernel/common",
    "branch": "android12-5.4",
    "commit": "de3036f041b742dfd2b0d86fc50ce61aaeca0aa1",
    "describe": "ASB-2025-06-05_12-5.4-2-gde3036f041b74",
    "version": {
     "version": 5,
     "patchlevel": 4,
     "sublevel": 292,
     "extra": "-5245-gde3036f041b74"
    },
    "commit_tags": [],
    "commit_message": "BACKPORT: ...",
    "tip_of_branch": true
   },
   "runtime": "k8s-all",
   "platform": "kubernetes",
   "job_id": "kci-6849fb29dba9af5ad62c489b-kbuild-clang-17-x86-andr-3xsdj1ze",
   "job_context": "gke_android-kernelci-external_us-east4-c_kci-big-us-east4",
   "error_code": "node_timeout",
   "error_msg": "Node timed-out"
  },
  "debug": {},
  "jobfilter": null,
  "platform_filter": [],
  "created": "2025-06-11T21:54:49.223000",
  "updated": "2025-06-12T04:00:58.200000",
  "timeout": "2025-06-12T03:54:49.223000",
  "holdoff": null,
  "owner": "production",
  "submitter": "service:pipeline",
  "treeid": "292f74a2cd242808aec687fdde0069e435a860ff4e00c53ed86cbbb0fb793ece",
  "user_groups": [],
  "processed_by_kcidb_bridge": true
 }
]
```

padovan commented:
> @aliceinwire I reformatted output to be nicer:
> 
> ```
> $ kci-dev --settings ./kci-dev.toml validate build --giturl https://android.googlesource.com/kernel/common --commit de3036f041b742dfd2b0d86fc50ce61aaeca0aa1 --branch android12-5.4
> Maestro:
> Build count:34
> Dashboard:
> Build count:33
> Build count didn't match
> [
>  {
>   "id": "6849fb29dba9af5ad62c489b",
>   "kind": "kbuild",
>   "name": "kbuild-clang-17-x86-android-allmodconfig",
>  ...
> ```

I believe that just putting a list of ids would be enough, or a list of id + link to our maestro api viewer, otherwise the output will be quite verbose.

aliceinwire commented:
> > @aliceinwire I reformatted output to be nicer:
> > ```
> > $ kci-dev --settings ./kci-dev.toml validate build --giturl https://android.googlesource.com/kernel/common --commit de3036f041b742dfd2b0d86fc50ce61aaeca0aa1 --branch android12-5.4
> > Maestro:
> > Build count:34
> > Dashboard:
> > Build count:33
> > Build count didn't match
> > [
> >  {
> >   "id": "6849fb29dba9af5ad62c489b",
> >   "kind": "kbuild",
> >   "name": "kbuild-clang-17-x86-android-allmodconfig",
> >  ...
> > ```
> 
> I believe that just putting a list of ids would be enough, or a list of id + link to our maestro api viewer, otherwise the output will be quite verbose.

Would be good to have by default a list of ids and have a `--verbose` argument for having the full verbose output

aliceinwire commented:
you added the commit [Enable --days switch for getting tree list](https://github.com/kernelci/kci-dev/pull/170/commits/a7b2f07bf11348731e1baba32f9190712db9dc69)[Enable --days switch for getting tree list](https://github.com/kernelci/kci-dev/pull/170/commits/a7b2f07bf11348731e1baba32f9190712db9dc69) of a different PR on this PR

JenySadadia commented:
> you added the commit [Enable --days switch for getting tree list](https://github.com/kernelci/kci-dev/pull/170/commits/a7b2f07bf11348731e1baba32f9190712db9dc69)[Enable --days switch for getting tree list](https://github.com/kernelci/kci-dev/pull/170/commits/a7b2f07bf11348731e1baba32f9190712db9dc69) of a different PR on this PR

Yes, this PR depends on 2 other PRs.
Otherwise, the command won't work properly. See https://github.com/kernelci/kci-dev/pull/170#issue-3120081972

padovan commented:
> The `git folder:None` print is from the existing utility I am using for fetching local repo if the options are not provided.
> See https://github.com/kernelci/kci-dev/blob/main/kcidev/libs/git_repo.py#L300
> We can tweak that function if the prints are misleading.

Yes, this should be fixed too.

JenySadadia commented:
> It looks good to me! Thanks for reworking it!

Okay. Thanks :)

JenySadadia commented:
@aliceinwire Could you please review the PR?

aliceinwire commented:
can you update this branch?
